ArduSub/control_motordetect.cpp

```diff
@@ -0,0 +1,175 @@
+#include "Sub.h"
+#include "stdio.h"
+
+/*
+ * control_motordetect.cpp - init and run calls for motordetect flightmode;
+ *
+ *  This mode pulses all thrusters to detect if they need to be reversed.
+ *  This still requires that the user has the correct frame selected and the motors
+ *  are connected to the correct ESCs.
+ *
+ *  For each motor:
+ *      wait until vehicle is stopped for > 500ms
+ *      apply throttle up for 500ms
+ *      If results are good:
+ *          save direction and try the next motor.
+ *      else
+ *          wait until vehicle is stopped for > 500ms
+ *          apply throttle down for 500ms
+ *          If results are good
+ *              save direction and try the next motor.
+ *          If results are bad
+ *              Abort!
+ */
+
+// controller states
+enum test_state {
+    STANDBY,
+    SETTLING,
+    THRUSTING,
+    DETECTING,
+    DONE
+};
+
+enum direction {
+    UP = 1,
+    DOWN = -1
+};
+
+static uint32_t settling_timer;
+static uint32_t thrusting_timer;
+static uint8_t md_state;
+static uint8_t current_motor;
+static int16_t current_direction;
+
+
+bool Sub::motordetect_init()
+{
+    current_motor = 0;
+    md_state = STANDBY;
+    current_direction = UP;
+    return true;
+}
+
+void Sub::motordetect_run()
+{
+    // if not armed set throttle to zero and exit immediately
+    if (!motors.armed()) {
+        motors.set_desired_spool_state(AP_Motors::DesiredSpoolState::GROUND_IDLE);
+        // Force all motors to stop
+        for(uint8_t i=0; i < AP_MOTORS_MAX_NUM_MOTORS; i++) {
+            if (motors.motor_is_enabled(i)) {
+                motors.output_test_num(i, 1500);
+            }
+        }
+        md_state = DONE;
```

Comment:
looks like this should be STANDBY. Or if we arm while inside this mode, we go straight to Done, which calls disarm.

Reply:
fixed